javalib/src/main/scala/java/util/Collections.scala

```diff
@@ -0,0 +1,1131 @@
+package java.util
+
+import java.{lang => jl, io}
+
+import scala.annotation.tailrec
+import scala.language.implicitConversions
+import scala.collection.JavaConversions._
+
+object Collections {
+
+  import WrappedCollections._
+  import ImmutableCollections._
+  import CheckedCollections._
+
+  final lazy val EMPTY_SET: Set[_] = {
+    unmodifiableSet(
+      new AbstractSet[Any] with io.Serializable {
+        override def size(): Int = 0
+
+        override def iterator(): Iterator[Any] = emptyIterator[Any]
+      })
+  }
+
+  final lazy val EMPTY_LIST: List[_] = {
+    unmodifiableList(
+      new AbstractList[Any] with io.Serializable {
+        override def get(index: Int): Any =
+          throw new IndexOutOfBoundsException(index.toString)
+
+        override def size(): Int = 0
+      })
+  }
+
+  final lazy val EMPTY_MAP: Map[_, _] = {
+    unmodifiableMap[Any, Any](
+      new AbstractMap[Any, Any] with io.Serializable {
+        override def entrySet(): Set[Map.Entry[Any, Any]] =
+          EMPTY_SET.asInstanceOf[Set[Map.Entry[Any, Any]]]
+      })
+  }
+
+  def sort[T <: jl.Comparable[T]](list: List[T]): Unit =
+    sort(list, makeComparator[T])
+
+  def sort[T](list: List[T], c: Comparator[_ >: T]): Unit = {
+    val sortedList = list.sorted(c)
+    (0 until list.size).foreach(i => list(i) = sortedList(i))
+  }
+
+  def binarySearch[T](list: List[_ <: jl.Comparable[T]], key: T): Int =
+    binarySearchCore(list, (elem: Comparable[T]) => elem.compareTo(key))
+
+  def binarySearch[T](list: List[_ <: T], key: T, c: Comparator[_ >: T]): Int =
+    binarySearchCore(list, (elem: T) => c.compare(elem, key))
+
+  def reverse(list: List[_]): Unit = {
+    val _list = list.asInstanceOf[List[AnyRef]]
+    val it1 = _list.listIterator()
+    val it2 = _list.listIterator(_list.size)
+    for (i <- 0 until _list.size / 2) {
+      val tmp = it1.next()
+      it1.set(it2.previous())
+      it2.set(tmp)
+    }
+  }
+
+  def shuffle(list: List[_]): Unit =
+    shuffle(list, new Random)
+
+  def shuffle(list: List[_], rnd: Random): Unit = {
+    val _list = list.asInstanceOf[List[AnyRef]]
+    val scalaRnd = scala.util.Random.javaRandomToRandom(rnd)
+    val shuffledList = scalaRnd.shuffle(_list.toSeq)
+    (0 until list.size).foreach(i => _list(i) = shuffledList(i))
+  }
+
+  def swap(list: List[_], i: Int, j: Int): Unit =
+    swapCore(list, i, j, true)
+
+  def fill[T](list: List[_ >: T], obj: T): Unit = {
+    @tailrec
+    def fillNext(iter: ListIterator[_ >: T]): Unit = {
+      if (iter.hasNext) {
+        iter.next()
+        iter.set(obj)
+        fillNext(iter)
+      }
+    }
+    fillNext(list.listIterator)
```

Comment:
Consider using a `while` loop. The tailrec here is nothing else than an implementation of a `while` loop. (That's often the case when iterating on a mutable Iterator.)